clients/cobol-lsp-vscode-extension/src/provider/TarCopybookFileSystemProvider.ts

```diff
@@ -0,0 +1,176 @@
+      return await TarUtil.readTarFile(tarFileUri);
+    },
+    undefined,
+    (tarFileUri: vscode.Uri) => {
+      const { tarfsPath, dialect } = this.getDetailsFromTarUri(tarFileUri);
+      return `${tarfsPath}$$${dialect}`;
+    },
+  );
+
+  onDidChangeFile: vscode.Event<vscode.FileChangeEvent[]> =
+    new vscode.EventEmitter<vscode.FileChangeEvent[]>().event;
+  watch(_uri: unknown, _options: unknown): vscode.Disposable {
+    throw new Error("Method not implemented.");
+  }
+
+  /**
+   * `${TarFileContentProvider.SCHEME}://${tarFileUri.fsPath}?filePath=${evaluatedInternalPath}#${dialect}`
+   * @param uri
+   * @returns
+   */
+  async stat(uri: vscode.Uri): Promise<vscode.FileStat> {
+    const { tarfsPath, filePath } = this.getDetailsFromTarUri(uri);
+    const matchingFiles = await this.fetchMatchingFiles(
+      tarfsPath,
+      uri,
+      filePath,
+    );
+    if (matchingFiles && matchingFiles[0]) {
+      return matchingFiles[0].fileData.fileMetadata;
+    }
+    throw vscode.FileSystemError.FileNotFound();
+  }
+
+  /**
+   * `${TarFileContentProvider.SCHEME}://${tarFileUri.fsPath}?searchPath=${evaluatedInternalPath}#${dialect}`
+   * get list of all the files within a TAR file and search pattern
+   * @param uri
+   */
+  async readDirectory(uri: vscode.Uri): Promise<[string, vscode.FileType][]> {
+    const { tarfsPath, searchPath } = this.getDetailsFromTarUri(uri);
+    const tarContent = await this.tarCache.execute(vscode.Uri.file(tarfsPath));
+    if (!tarContent) {
+      throw new Error(`Failed to retrieve content for URI: ${uri.toString()}`);
+    }
+    const result: [string, vscode.FileType][] = [];
+    const matchingFiles = tarContent.filter((e) =>
+      new Minimatch(searchPath + "/*", {
+        nocase: true,
+        dot: true,
+      }).match(e.fileName),
+    );
+    matchingFiles.forEach((t) =>
+      result.push([t.fileName, t.fileData.fileMetadata.type]),
+    );
+    return result;
+  }
+
+  createDirectory(_uri: unknown): void | Thenable<void> {
+    throw new Error("Method not implemented.");
+  }
+
+  /**
+   * `${TarFileContentProvider.SCHEME}://${tarFileUri.fsPath}?filePath=${evaluatedInternalPath}#${dialect}`
+   * @param uri
+   * @returns
+   */
+  async readFile(uri: vscode.Uri) {
+    const { tarfsPath, filePath, dialect } = this.getDetailsFromTarUri(uri);
+
+    const matchingFiles = await this.fetchMatchingFiles(
+      tarfsPath,
+      uri,
+      filePath,
+    );
+
+    if (
+      matchingFiles &&
+      matchingFiles.length > 0 &&
+      matchingFiles[0].fileData &&
+      matchingFiles[0].fileData.fileContent instanceof Uint8Array
+    ) {
+      // return the first found file
+      return matchingFiles[0].fileData.fileContent;
+    }
+    console.log(
+      `file ${filePath} not found in tar ${tarfsPath} for dialect ${dialect}`,
+    );
+    throw vscode.FileSystemError.FileNotFound();
+  }
+
+  /**
+   * clear cache
+   */
+  public clearCache() {
+    this.tarCache.clearCache();
```

Comment:
We should "delete" the top level directories `onDidChangeFile`.

Reply:
would this works ? 2f1df14b23739f2a58a46792dc169e4f3aac943f